bauble_macro_util/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+use std::collections::HashSet;
+
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::quote;
+use syn::{
+    parse2, spanned::Spanned, AttrStyle, Attribute, Data, DeriveInput, Error, Expr, Fields,
+    ImplGenerics, Index, Token, Type, WhereClause,
+};
+
+// Related fields used by `derive_struct` and `derive_fields` containing type info
+struct TypeInfo<'a> {
+    // The struct or variant, used for construction
+    ty: proc_macro2::TokenStream,
+    // The type's generics
+    impl_generics: &'a ImplGenerics<'a>,
+    where_clause: Option<&'a WhereClause>,
+}
+
+// Generate code to deserialize a struct or variant with fields
+fn derive_fields(
+    TypeInfo {
+        ty,
+        impl_generics,
+        where_clause,
+    }: TypeInfo,
+    // The struct or variant's fields
+    fields: &Fields,
+    // Whether the struct or variant should be parsed from a tuple. For structs with named
+    // fields, this is the case if it has the `tuple` attribute
+    tuple: bool,
+) -> proc_macro2::TokenStream {
+    // General kind of field
+    enum FieldTy<'a> {
+        // The field may be deserialized from `bauble`, and must implement `FromBauble`
+        Val {
+            // An expression to generate this type. If `Some`, the field does not need to be
+            // specified in `bauble`.
+            default: Option<proc_macro2::TokenStream>,
+            // Whether the field is a `bauble` attribute
+            attribute: bool,
+            // Index for a tuple that holds the values of deserializable fields
+            index: Index,
+            // Type from which the field is deserialized
+            ty: &'a Type,
+        },
+        // The field is only generated from a default expression
+        AsDefault {
+            // An expression to generate this type
+            default: proc_macro2::TokenStream,
+            // Type from which the field is deserialized
+            ty: &'a Type,
+        },
+    }
+
+    // Information about a field collected from its attributes
+    struct FieldAttrs<'a> {
+        name: proc_macro2::TokenStream,
+        ty: FieldTy<'a>,
+    }
+
+    let mut val_count = 0;
+
+    // Parse the fields and attributes
+    let fields = match fields
+        .iter()
+        .enumerate()
+        .map(|(index, field)| -> Result<_, proc_macro2::TokenStream> {
```

Comment:
I checked, and the compiler can infer the return type of this closure without the type hint.